### PR TITLE
1653 jr-filtering-5-provider-and-or-brand-level-job-group-filter

### PR DESCRIPTION
**Pull request by MarionJHolloway**
## Description
Trello ticket [#1653](https://trello.com/c/I6BPziY7/1653-jr-filtering-5-provider-and-or-brand-level-job-group-filter)

Refactor job group filter to work at brand/ provider or location level

## Testing
- [x] Unit tests passing
- [x] Successful [Step Function run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:856699698263:execution:1653-jr-fil-5v3-Ind-CQC-Filled-Post-Estimates-By-Role:47e8036f-b4c5-415a-a453-250133f32cf5)
- [x] Outputs checked in Athena

Filtered outputs in ticket (~34,000 rows)

## Checklist (delete if not relevant)
- [x] Unit tests added/amended
- [x] Docstrings added/updated
- [x] Updated CHANGELOG
- [x] Code reviewed by AI
- [x] Moved Trello ticket to PR column

## Reviewer checklist
- [ ] PR solves the Trello ticket requirements
- [ ] Outputs appear reasonable and understandable
- [ ] The overall approach is appropriate and not over-engineered
- [ ] No obvious scalability, performance or interdependency concerns
- [ ] Tests appropriately cover the main behaviour/edge cases
- [ ] Naming and structure are broadly understandable and consistent
- [ ] Docstrings are sufficient for future maintenance
- [ ] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py [hunk 1]
```diff
@@ -98,81 +103,109 @@ def filter_job_role_group_outliers(
             values=IndCQC.ascwds_job_role_counts,
             aggregate_function="sum",
         )
-        .with_columns(Exprs.location_sum_expr)
-        .filter(pl.col(Exprs.temp_location_sum) > small_location_threshold)
+        .with_columns(Exprs.id_column_sum_expr)
+        .filter(pl.col(Exprs.temp_id_column_sum) > min_workers_threshold)
```
Comment (AchyutRatkanthiwar) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py]: The doc string says "same or more workers than the min workers threshold", so should this be >= ?

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py [hunk 2]
```diff
@@ -98,81 +103,109 @@ def filter_job_role_group_outliers(
...
+    job_group_bounds_cols: list[str]
+    job_role_group_bounds_locations: dict[str, dict[str, float]]
+    job_role_group_bounds_brand_prov: dict[str, float]
+    id_column_sum_expr: pl.Expr
     job_group_percentage_expr: pl.Expr
     evaluation_expr: pl.Expr
 
     def __init__(self):
-        self.temp_location_sum = "location_sum"
+        self.temp_id_column_sum = "location_sum"
```
Comment (AchyutRatkanthiwar) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py]: This should be self.temp_id_column_sum = "id_column_sum"

File: projects/_03_independent_cqc/unittest_data/polars_ind_cqc_test_file_schemas.py [hunk 3]
```diff
@@ -1715,19 +1713,65 @@ class EstimateFilledPostsByJobRoleCleanUtilsSchemas:
         IndCQC.job_role_filtering_rule: CatColType.JobRoleFilteringRuleCatType,
         IndCQC.job_role_filtering_rule: CatColType.JobRoleFilteringRuleCatType,
```
Comment (AchyutRatkanthiwar) [inline projects/_03_independent_cqc/unittest_data/polars_ind_cqc_test_file_schemas.py]: I these are repeated. not sure why running tests didnt flag these

File: projects/_03_independent_cqc/unittest_data/polars_ind_cqc_test_file_schemas.py [hunk 4]
```diff
@@ -1715,19 +1713,65 @@ class EstimateFilledPostsByJobRoleCleanUtilsSchemas:
...
+        IndCQC.provider_id: pl.String,
+        IndCQC.cqc_location_import_date: pl.Date,
+        IndCQC.primary_service_type: CatColType.PrimaryServiceEnumType,
+        IndCQC.main_job_role_clean_labelled: CatColType.JobRoleEnumType,
+        IndCQC.main_job_group_labelled: pl.Enum(
+            AscwdsWorkerValueLabelsJobGroup.all_job_groups()
+        ),
+        IndCQC.ascwds_job_role_counts: pl.Int64,
+        IndCQC.job_role_filtering_rule: CatColType.JobRoleFilteringRuleCatType,
+        IndCQC.job_role_filtering_rule: CatColType.JobRoleFilteringRuleCatType,
```
Comment (AchyutRatkanthiwar) [inline projects/_03_independent_cqc/unittest_data/polars_ind_cqc_test_file_schemas.py]: I these are repeated. not sure why running tests didnt flag these

File: projects/_03_independent_cqc/unittest_data/polars_ind_cqc_test_file_schemas.py [hunk 5]
```diff
@@ -1715,19 +1713,65 @@ class EstimateFilledPostsByJobRoleCleanUtilsSchemas:
...
+        IndCQC.brand_id: pl.String,
+        IndCQC.cqc_location_import_date: pl.Date,
+        IndCQC.primary_service_type: CatColType.PrimaryServiceEnumType,
+        IndCQC.main_job_role_clean_labelled: CatColType.JobRoleEnumType,
+        IndCQC.main_job_group_labelled: pl.Enum(
+            AscwdsWorkerValueLabelsJobGroup.all_job_groups()
+        ),
+        IndCQC.ascwds_job_role_counts: pl.Int64,
+        IndCQC.job_role_filtering_rule: CatColType.JobRoleFilteringRuleCatType,
+        IndCQC.job_role_filtering_rule: CatColType.JobRoleFilteringRuleCatType,
```
Comment (AchyutRatkanthiwar) [inline projects/_03_independent_cqc/unittest_data/polars_ind_cqc_test_file_schemas.py]: I these are repeated. not sure why running tests didnt flag these

**Review by AchyutRatkanthiwar — COMMENTED** [review]
looks good. Just some small changes in schema and docstring

File: projects/_03_independent_cqc/unittest_data/polars_ind_cqc_test_file_schemas.py (hunk 5 above)
Comment (MarionJHolloway) [reply projects/_03_independent_cqc/unittest_data/polars_ind_cqc_test_file_schemas.py]: I deleted these once already! Worth checking again before merging.

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py [hunk 6]
```diff
@@ -46,50 +46,55 @@ def nullify_job_role_count_when_source_not_ascwds(lf: pl.LazyFrame) -> pl.LazyFr
 
 def filter_job_role_group_outliers(
```
Comment (garypolzin) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py:47]: this possibly isn't worth the time refactoring but I think it can be a lot simpler if it's done in one go, rather than 3 loops through the same function with all the if/else statements

Create lookup lazyframe
- define the magic numbers in a lazyframe instead of a dict (service and jobrolegroup as cols, lower and upper as cols
- calculate max of those services to get brand/prov limits (saves defining even more magic numbers)

Join in to main lazyframe matched on service and job group

calculate brand/provider totals/proportions (as cols or as expressions)

then it's a simple comparison of
    null if brand out of bounds
    update rule
    null if provider out of bounds
    update rule
    null if location out of bounds
    update rule

Comment (garypolzin) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py:47]: can probably make that more efficient with expressions - don't actually need to calculate the sums and proportions as physical columns

Comment (MarionJHolloway) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py:47]: Discussed with Gary, decided probably not worth the work atm but might be worth it if the polars over function ever comes to streaming.

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py [hunk 7]
```diff
@@ -46,50 +46,56 @@ def nullify_job_role_count_when_source_not_ascwds(lf: pl.LazyFrame) -> pl.LazyFr
...
     - managers, regulated professions, or other are outside their upper bound
 
     Args:
         lf (pl.LazyFrame): The estimated filled post by job role LazyFrame.
-        small_location_threshold (int): Minimum number of workers at a location to
-            be included in filtering. Defaults to 50.
+        id_column (str): The id column on which the filter should be applied. Possible
+            values: location id, provider id, brand id. Defaults to location id.
+        min_workers_threshold (int): Minimum number of workers at a location/ provider/
+            brand to be included in filtering. If below this treshold, location/ provider/brand
```
Comment (AchyutRatkanthiwar) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/clean_utils.py]: threshold

**Review by AchyutRatkanthiwar — APPROVED** [review]
